eZ/Publish/Core/Repository/TrashService.php

```diff
@@ -171,7 +173,12 @@ public function recover(APITrashItem $trashItem, Location $newParentLocation = n
             throw new InvalidArgumentValue('parentLocationId', $newParentLocation->id, 'Location');
         }
 
-        if ($this->repository->hasAccess('content', 'restore') !== true) {
+        if (!$this->repository->canUser(
+            'content',
+            'restore',
+            $trashItem->getContentInfo(),
```

Comment:
The same comment as above, but I guess we just need to check this when moved to QA.

Reply:
Yes, this one needs to be tested, the change implies role limitation will kick in here only allowing you to restore within trees you have been granted role with restore rights.

Which I think makes sense here as we have a proper location which is a target of the action _(both in case different location was provided and in case where trash location is to be the target)_


As for "above", this is not the case anymore for checks on read permission:
1. It does now not include the trash location there _(as it is about checking if you can read content in trash, location is irrelevant)_
2. Changed SubtreeLimitationType to abstain if no target is provided and content is in trash to make this work.